contracts/misc/TimeLock.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import {IERC20} from "../dependencies/openzeppelin/contracts/IERC20.sol";
+import {IERC721} from "../dependencies/openzeppelin/contracts/IERC721.sol";
+import {IERC1155} from "../dependencies/openzeppelin/contracts/IERC1155.sol";
+import "../dependencies/openzeppelin/upgradeability/OwnableUpgradeable.sol";
+import "../dependencies/openzeppelin/upgradeability/ReentrancyGuardUpgradeable.sol";
+import {EnumerableSet} from "../dependencies/openzeppelin/contracts/EnumerableSet.sol";
+import {ITimeLock} from "../interfaces/ITimeLock.sol";
+import {IPool} from "../interfaces/IPool.sol";
+import {DataTypes} from "../protocol/libraries/types/DataTypes.sol";
+import {GPv2SafeERC20} from "../dependencies/gnosis/contracts/GPv2SafeERC20.sol";
+
+contract TimeLock is ITimeLock, OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using GPv2SafeERC20 for IERC20;
+
+    struct Agreement {
+        DataTypes.AssetType assetType;
+        DataTypes.TimeLockActionType actionType;
+        bool isFrozen;
+        address asset;
+        address beneficiary;
+        uint48 releaseTime;
+        uint256[] tokenIdsOrAmounts;
+    }
+
+    event AgreementCreated(
+        uint256 agreementId,
+        DataTypes.AssetType assetType,
+        address indexed asset,
+        uint256[] tokenIdsOrAmounts,
+        address indexed beneficiary,
+        uint48 releaseTime
+    );
+
+    event AgreementClaimed(
+        uint256 agreementId,
+        DataTypes.AssetType assetType,
+        address indexed asset,
+        uint256[] tokenIdsOrAmounts,
+        address indexed beneficiary
+    );
+    mapping(uint256 => Agreement) public agreements;
```

Comment:
@WalidOfNow The public view function `agreements(id)` just returns (assetType, actionType, isFrozen, asset, beneficiary, releaseTime), not contains `tokenIdsOrAmounts `, you need add another view function for us to get the  
 full Agreement data.

Reply:
View function can be added tomorrow.  Let's have the PR ready for secure3 first I think

Reply:
Yeah, I agree.